### PR TITLE
Handle `-inf` and `inf` values in `RDBStorage`

### LINKED ISSUE #3206: `inf` handling in Optuna
## Expected behavior

Regardless of the type of storage used, the trial accepts `inf` for its value when an objective function returns `inf`, when `inf` is passed in `trial.report`, or when `inf` is passed in `study.tell`. Note that it is the responsibility of the sampler or pruner to decide how to use the `inf` values stored in the storage. For example, `TPESampler`, the default sampler in Optuna, is implemented to be able to use `inf`.

We need to fix the followings.
- [x] If `inf` is returned by objective function, the study accepts it regardless of the type of storage used.
- [x] If `inf` is passed in `trial.report`, the study accepts it regardless of the type of storage used.
- [x] If `inf` is passed in `study.tell`, the study accepts it regardless of the type of storage used.

## Environment

- Optuna version: a70cbff8dbc454f137b61622633404bd4e4657fc
- Python version: 3.9.9
- OS: macOS Catalina 10.15.7
- (Optional) Other libraries and their versions:
psycopg2:    2.9.2
PyMySQL:     1.0.2

## Error messages, stack traces, or logs

If we run the script `inf-objective.py`, we will get the following output. The script is provided in the section of `Reproducible examples`. I think we need to modify this behavior as explained in `Expected behavior`. Concretely, we should fix the behavior with MySQL to accept the `inf`.
<details>
<summary>output</summary>

```
(venv) mamu@HideakinoMacBook-puro test % python inf-objective.py 
[I 2021-12-24 13:58:54,771] A new study created in memory with name: no-name-f7f28497-2f1c-4739-b9c8-edca726274fc
[I 2021-12-24 13:58:54,772] Trial 0 finished with value: inf and parameters: {}. Best is trial 0 with value: inf.
In-memory : Trial state is TrialState.COMPLETE
[I 2021-12-24 13:58:54,843] A new study created in RDB with name: no-name-168f23c0-0ac5-4785-a6d6-bdfd31dda5ab
[I 2021-12-24 13:58:54,909] Trial 0 finished with value: inf and parameters: {}. Best is trial 0 with value: inf.
SQLite    : Trial state is TrialState.COMPLETE
[I 2021-12-24 13:58:55,052] A new study created in RDB with name: no-name-22068297-6685-4435-bf88-124621f797c1
Record does not exist.
MySQL     : Trial state is TrialState.RUNNING
[I 2021-12-24 13:58:55,501] A new study created in RDB with name: no-name-39229033-d3b5-43fa-93f6-1e9a313593f6
[I 2021-12-24 13:58:55,700] Trial 0 finished with value: inf and parameters: {}. Best is trial 0 with value: inf.
PostgreSQL: Trial state is TrialState.COMPLETE
```
</details>

On the other hand, if we run the script `inf-report.py`, we will get the following output. The script is provided in the section of `Reproducible examples`. I think we need to modify this behavior as explained in `Expected behavior`. Concretely, we should fix the behavior with MySQL to accept the `inf`.
<details>
<summary>output</summary>

```
(venv) mamu@HideakinoMacBook-puro test % python inf-report.py 
[I 2021-12-24 14:01:01,970] A new study created in memory with name: no-name-44b8d7da-1b6f-4e38-80f4-33d826b4ff74
[I 2021-12-24 14:01:01,971] Trial 0 finished with value: 1.0 and parameters: {}. Best is trial 0 with value: 1.0.
In-memory : Trial state is TrialState.COMPLETE
[I 2021-12-24 14:01:02,040] A new study created in RDB with name: no-name-8e47d68f-83e3-4016-b18c-b7ae4930986b
[I 2021-12-24 14:01:02,564] Trial 0 finished with value: 1.0 and parameters: {}. Best is trial 0 with value: 1.0.
SQLite    : Trial state is TrialState.COMPLETE
[I 2021-12-24 14:01:02,712] A new study created in RDB with name: no-name-dede90c7-9907-47d5-9a78-873233f61f36
[W 2021-12-24 14:01:02,830] Trial 0 failed because of the following error: StorageInternalError('An exception is raised during the commit. This typically happens due to invalid data in the commit, e.g. exceeding max length. ')
Traceback (most recent call last):
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1802, in _execute_context
    self.dialect.do_execute(
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 732, in do_execute
    cursor.execute(statement, parameters)
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/pymysql/cursors.py", line 146, in execute
    query = self.mogrify(query, args)
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/pymysql/cursors.py", line 125, in mogrify
    query = query % self._escape_args(args, conn)
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/pymysql/cursors.py", line 109, in _escape_args
    return {key: conn.literal(val) for (key, val) in args.items()}
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/pymysql/cursors.py", line 109, in <dictcomp>
    return {key: conn.literal(val) for (key, val) in args.items()}
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/pymysql/connections.py", line 517, in literal
    return self.escape(obj, self.encoders)
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/pymysql/connections.py", line 510, in escape
    return converters.escape_item(obj, self.charset, mapping=mapping)
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/pymysql/converters.py", line 25, in escape_item
    val = encoder(val, mapping)
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/pymysql/converters.py", line 60, in escape_float
    raise ProgrammingError("%s can not be used with MySQL" % s)
pymysql.err.ProgrammingError: inf can not be used with MySQL

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/optuna/storages/_rdb/storage.py", line 54, in _create_scoped_session
    session.commit()
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 1431, in commit
    self._transaction.commit(_to_root=self.future)
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 829, in commit
    self._prepare_impl()
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 808, in _prepare_impl
    self.session.flush()
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 3363, in flush
    self._flush(objects)
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 3503, in _flush
    transaction.rollback(_capture_exception=True)
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 3463, in _flush
    flush_context.execute()
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/sqlalchemy/orm/unitofwork.py", line 456, in execute
    rec.execute(self)
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/sqlalchemy/orm/unitofwork.py", line 630, in execute
    util.preloaded.orm_persistence.save_obj(
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/sqlalchemy/orm/persistence.py", line 244, in save_obj
    _emit_insert_statements(
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/sqlalchemy/orm/persistence.py", line 1221, in _emit_insert_statements
    result = connection._execute_20(
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1614, in _execute_20
    return meth(self, args_10style, kwargs_10style, execution_options)
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 325, in _execute_on_connection
    return connection._execute_clauseelement(
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1481, in _execute_clauseelement
    ret = self._execute_context(
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1845, in _execute_context
    self._handle_dbapi_exception(
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2026, in _handle_dbapi_exception
    util.raise_(
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1802, in _execute_context
    self.dialect.do_execute(
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 732, in do_execute
    cursor.execute(statement, parameters)
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/pymysql/cursors.py", line 146, in execute
    query = self.mogrify(query, args)
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/pymysql/cursors.py", line 125, in mogrify
    query = query % self._escape_args(args, conn)
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/pymysql/cursors.py", line 109, in _escape_args
    return {key: conn.literal(val) for (key, val) in args.items()}
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/pymysql/cursors.py", line 109, in <dictcomp>
    return {key: conn.literal(val) for (key, val) in args.items()}
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/pymysql/connections.py", line 517, in literal
    return self.escape(obj, self.encoders)
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/pymysql/connections.py", line 510, in escape
    return converters.escape_item(obj, self.charset, mapping=mapping)
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/pymysql/converters.py", line 25, in escape_item
    val = encoder(val, mapping)
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/pymysql/converters.py", line 60, in escape_float
    raise ProgrammingError("%s can not be used with MySQL" % s)
sqlalchemy.exc.ProgrammingError: (pymysql.err.ProgrammingError) inf can not be used with MySQL
[SQL: INSERT INTO trial_intermediate_values (trial_id, step, intermediate_value) VALUES (%(trial_id)s, %(step)s, %(intermediate_value)s)]
[parameters: {'trial_id': 10, 'step': 1, 'intermediate_value': inf}]
(Background on this error at: https://sqlalche.me/e/14/f405)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/optuna/study/_optimize.py", line 213, in _run_trial
    value_or_values = func(trial)
  File "/Users/mamu/Dev/pfn/code_snippets_for_optuna/test/inf-report.py", line 5, in objective
    trial.report(float(np.inf), 1)
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/optuna/trial/_trial.py", line 506, in report
    self.storage.set_trial_intermediate_value(self._trial_id, step, value)
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/optuna/storages/_cached_storage.py", line 293, in set_trial_intermediate_value
    self._backend.set_trial_intermediate_value(trial_id, step, intermediate_value)
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/optuna/storages/_rdb/storage.py", line 754, in set_trial_intermediate_value
    self._set_trial_intermediate_value_without_commit(
  File "/Users/mamu/.pyenv/versions/3.9.9/lib/python3.9/contextlib.py", line 126, in __exit__
    next(self.gen)
  File "/Users/mamu/Dev/pfn/venv/lib/python3.9/site-packages/optuna/storages/_rdb/storage.py", line 71, in _create_scoped_session
    raise optuna.exceptions.StorageInternalError(message) from e
optuna.exceptions.StorageInternalError: An exception is raised during the commit. This typically happens due to invalid data in the commit, e.g. exceeding max length. 
An exception is raised during the commit. This typically happens due to invalid data in the commit, e.g. exceeding max length. 
MySQL     : Trial state is TrialState.FAIL
[I 2021-12-24 14:01:03,078] A new study created in RDB with name: no-name-da490862-bbd5-4e04-8b64-be71deddeb52
[I 2021-12-24 14:01:03,238] Trial 0 finished with value: 1.0 and parameters: {}. Best is trial 0 with value: 1.0.
PostgreSQL: Trial state is TrialState.COMPLETE
```
</details>

In addition, if we run the script `inf-tell.py`, we will get the following output. The script is provided in the section of `Reproducible examples`. I think we need to modify this behavior as explained in `Expected behavior`. Concretely, we should fix the behavior with MySQL to accept the `inf`.
<details>
<summary>output</summary>

```
[I 2021-12-24 15:26:11,042] A new study created in memory with name: no-name-7cc77d37-ffee-45a6-8488-8b66aeda30e8
In-memory : Trial state is TrialState.COMPLETE
[I 2021-12-24 15:26:11,105] A new study created in RDB with name: no-name-5200946f-3b52-4043-b477-74c84d1fa160
SQLite    : Trial state is TrialState.COMPLETE
[I 2021-12-24 15:26:11,287] A new study created in RDB with name: no-name-5821e24f-163e-4e3b-9702-6adb73fe2311
An exception is raised during the commit. This typically happens due to invalid data in the commit, e.g. exceeding max length. 
MySQL     : Trial state is TrialState.RUNNING
[I 2021-12-24 15:26:11,613] A new study created in RDB with name: no-name-218980c1-e3ff-48b8-b349-243b210034f7
PostgreSQL: Trial state is TrialState.COMPLETE
```
</details>

## Steps to reproduce

1. Install docker
2. Setup MySQL DB
```
docker run --name mysql -e MYSQL_ROOT_PASSWORD=test -p 3306:3306 -p 33060:33060 -d mysql:8.0.22
docker run --network host -it --rm mysql:8.0.22  mysql -h 127.0.0.1 -uroot -ptest -e "create database mysql;"
```
3. Setup PostgreSQL DB
```
docker run -it --rm --name postgres-test -e POSTGRES_PASSWORD=test -p 15432:5432 -d postgres
```
4. Run `inf-objectibe.py`, `inf-report.py`, and `inf-tell.py`.

## Reproducible examples (optional)
- `inf-objective.py`
```python
import optuna
import numpy as np

def objective(trial):
    return float(np.inf)

# In-memory storage
study = optuna.create_study()
study.optimize(objective, n_trials=1)
print(f"In-memory : Trial state is {study.trials[-1].state}")

# RDB storage (SQLite)
study = optuna.create_study(storage='sqlite:///../sqlite.db')
study.optimize(objective, n_trials=1)
print(f"SQLite    : Trial state is {study.trials[-1].state}")

# RDB storage (MySQL)
study = optuna.create_study(storage="mysql+pymysql://root:test@localhost/mysql")
try:
    study.optimize(objective, n_trials=1)
except Exception as e:
    print(e)
    print(f"MySQL     : Trial state is {study.trials[-1].state}")

# RDB storage (PostgreSQL)
study = optuna.create_study(storage="postgresql+psycopg2://postgres:test@localhost:15432/postgres")
study.optimize(objective, n_trials=1)
print(f"PostgreSQL: Trial state is {study.trials[-1].state}")
```

- `inf-report.py`
```python
import optuna
import numpy as np

def objective(trial):
    trial.report(float(np.inf), 1)
    return 1

# In-memory storage
study = optuna.create_study()
study.optimize(objective, n_trials=1)
print(f"In-memory : Trial state is {study.trials[-1].state}")

# RDB storage (SQLite)
study = optuna.create_study(storage='sqlite:///../sqlite.db')
study.optimize(objective, n_trials=1)
print(f"SQLite    : Trial state is {study.trials[-1].state}")

# RDB storage (MySQL)
study = optuna.create_study(storage="mysql+pymysql://root:test@localhost/mysql")
try:
    study.optimize(objective, n_trials=1)
except Exception as e:
    print(e)
    print(f"MySQL     : Trial state is {study.trials[-1].state}")

# RDB storage (PostgreSQL)
study = optuna.create_study(storage="postgresql+psycopg2://postgres:test@localhost:15432/postgres")
study.optimize(objective, n_trials=1)
print(f"PostgreSQL: Trial state is {study.trials[-1].state}")
``` 

- `inf-tel.py`
```python
import optuna
import numpy as np


# In-memory storage
study = optuna.create_study()
trial = study.ask()
study.tell(trial, float(np.inf))
print(f"In-memory : Trial state is {study.trials[-1].state}")

# RDB storage (SQLite)
study = optuna.create_study(storage='sqlite:///../sqlite.db')
trial = study.ask()
study.tell(trial, float(np.inf))
print(f"SQLite    : Trial state is {study.trials[-1].state}")

# RDB storage (MySQL)
study = optuna.create_study(storage="mysql+pymysql://root:test@localhost/mysql")
trial = study.ask()
try:
    study.tell(trial, float(np.inf))
except Exception as e:
    print(e)
    print(f"MySQL     : Trial state is {study.trials[-1].state}")

# RDB storage (PostgreSQL)
study = optuna.create_study(storage="postgresql+psycopg2://postgres:test@localhost:15432/postgres")
trial = study.ask()
study.tell(trial, float(np.inf))
print(f"PostgreSQL: Trial state is {study.trials[-1].state}")
```

### PR DESCRIPTION
**Pull request by xadrianzetx**
## Motivation

Previously an error could be raised when `-inf` or `inf` was registered in rdb storage. Depending on SQL dialect in backend, those special values might be supported or not. This PR introduces a simple way to get around this limitation by storing `-inf` and `inf` as minimum and maximum values of 32 bit float and providing thin translation layer between stored and true values on database reads and writes. This gives effective available range for trial values of `(-3.4028235e+38, 3.4028235e+38)`. Everything beyond that is interpreted as infinity. Closes #3206.

## Description of the changes

* Implement `inf` handling in `RDBStorage`
* Test

### CODE REVIEW

**Comment by xadrianzetx** [comment]
Note - this is a simple way of fixing it without need to modify data models. The downside is that when translating value from storage back to actual value, we can't directly compare two floats, only approximations. This means some trial values that are close enough to numerical limits might be translated back as infinite. But still, this is for values close to `2e+38` so in reality this is probably a non issue.

Alternative solution is to modify `TrialValueModel` and `TrialIntermediateValueModel` to include an integer column which indicates if value is infinite. This would also give us a chance to specify double precision for `Float` in those two models, meaning reported value could be in  `(-1.7976931348623157e+308, 1.7976931348623157e+308)` range and would go in line with python float range (which is C type `double` underneath).

**Comment by toshihikoyanase** [comment]
@nzw0301 @HideakiImamura Could you review this PR, please?

**Comment by nzw0301** [comment]
Sure. I'll review this PR within the next week.

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR. Basically, LGTM. Let me add a comment for the test.

File: tests/storages_tests/rdb_tests/test_with_server.py [hunk 1]
```diff
@@ -102,6 +102,26 @@ def test_loaded_trials(storage_url: str) -> None:
     _check_trials(loaded_study.trials)
 
 
+@pytest.mark.parametrize(
+    "input,expected",
+    [
+        (float("inf"), float("inf")),
+        (-float("inf"), -float("inf")),
+        (np.finfo(np.float32).max, float("inf")),
+        (np.finfo(np.float32).min, -float("inf")),
```
[the remark this answers is not in the document]
Comment (xadrianzetx) [reply tests/storages_tests/rdb_tests/test_with_server.py:111]: Done in cf8a45c80dfba8b0c0b3a8b0bfa177f7dd482682.

**Review by HideakiImamura — APPROVED** [review]
Thanks for the update! LGTM.

**Comment by HideakiImamura** [comment]
I confirmed that all problems in #3206 are resolved by this PR.
```
(venv) mamu@HideakinoMacBook-puro 3205-3206 % python inf-objective.py
[I 2022-01-28 12:56:37,375] A new study created in memory with name: no-name-ff107e9e-301c-460b-b9cc-81df43d3968d
[I 2022-01-28 12:56:37,376] Trial 0 finished with value: inf and parameters: {}. Best is trial 0 with value: inf.
In-memory : Trial state is TrialState.COMPLETE
[I 2022-01-28 12:56:37,441] A new study created in RDB with name: no-name-969966ff-fb18-494a-932c-ecddc8a4017b
[I 2022-01-28 12:56:37,502] Trial 0 finished with value: inf and parameters: {}. Best is trial 0 with value: inf.
SQLite    : Trial state is TrialState.COMPLETE
[I 2022-01-28 12:56:37,620] A new study created in RDB with name: no-name-07b13b5f-85bc-46a2-ba72-8b95dbb7eb5f
[I 2022-01-28 12:56:37,763] Trial 0 finished with value: inf and parameters: {}. Best is trial 0 with value: inf.
MySQL     : Trial state is TrialState.COMPLETE
[I 2022-01-28 12:56:37,903] A new study created in RDB with name: no-name-90b645ef-80ff-4730-a1e0-01e2c51011c5
[I 2022-01-28 12:56:38,012] Trial 0 finished with value: inf and parameters: {}. Best is trial 0 with value: inf.
PostgreSQL: Trial state is TrialState.COMPLETE
(venv) mamu@HideakinoMacBook-puro 3205-3206 % python inf-report.py   
[I 2022-01-28 12:56:41,181] A new study created in memory with name: no-name-d9311317-b070-46ea-a22d-39c280b29046
[I 2022-01-28 12:56:41,182] Trial 0 finished with value: 1.0 and parameters: {}. Best is trial 0 with value: 1.0.
In-memory : Trial state is TrialState.COMPLETE
[I 2022-01-28 12:56:41,247] A new study created in RDB with name: no-name-13547701-5097-4ca9-b34a-a7dc1810e732
[I 2022-01-28 12:56:41,315] Trial 0 finished with value: 1.0 and parameters: {}. Best is trial 0 with value: 1.0.
SQLite    : Trial state is TrialState.COMPLETE
[I 2022-01-28 12:56:41,435] A new study created in RDB with name: no-name-c383b025-e13b-4122-9f23-0e1c2b4680ae
[I 2022-01-28 12:56:41,575] Trial 0 finished with value: 1.0 and parameters: {}. Best is trial 0 with value: 1.0.
MySQL     : Trial state is TrialState.COMPLETE
[I 2022-01-28 12:56:41,712] A new study created in RDB with name: no-name-21630f5e-9773-4e9a-bbf9-8011f1c81a53
[I 2022-01-28 12:56:41,829] Trial 0 finished with value: 1.0 and parameters: {}. Best is trial 0 with value: 1.0.
PostgreSQL: Trial state is TrialState.COMPLETE
(venv) mamu@HideakinoMacBook-puro 3205-3206 % python inf-tell.py 
[I 2022-01-28 12:56:47,353] A new study created in memory with name: no-name-66267843-18bf-4c88-8701-2643a782c1d6
In-memory : Trial state is TrialState.COMPLETE
[I 2022-01-28 12:56:47,420] A new study created in RDB with name: no-name-b8cd0fce-6807-4174-8606-b352d214b19d
SQLite    : Trial state is TrialState.COMPLETE
[I 2022-01-28 12:56:47,583] A new study created in RDB with name: no-name-db63b7aa-e06f-4d1d-b67c-1bdc7abe8268
MySQL     : Trial state is TrialState.COMPLETE
[I 2022-01-28 12:56:47,806] A new study created in RDB with name: no-name-121048c0-3d33-408b-b56f-103e83c06e06
PostgreSQL: Trial state is TrialState.COMPLETE
```

**Review by nzw0301 — APPROVED** [review]
Thank you for your PR! LGTM. 

I locally confirmed the code works with 
1. a single objective (`inf`) with intermediate values (`inf`)
2. two objective values (`inf, 1.0`).


I'm wondering if we could add the explanation of this transformation to Optuna documentation, where `storage` or `study`. But it can be addressed as a follow-up.

So I'll merge this PR once.

**Pull request merged**